include/zephyr/device.h

```diff
@@ -923,7 +926,634 @@ BUILD_ASSERT(sizeof(device_handle_t) == 2, "fix the linker scripts");
 	Z_INIT_ENTRY_DEFINE(DEVICE_NAME_GET(dev_name), init_fn,		\
 		(&DEVICE_NAME_GET(dev_name)), level, prio)
 
-#if CONFIG_HAS_DTS
+/**
+ * @brief Multi API Device Model APIs
+ * @defgroup device_model_multi_api Multi API Device Model APIs
+ *
+ * @note The multi API devices do not support device handles, and can not
```

Comment:
Oh, I am afraid that's a huge no-go.

You have to find a way to keep all the devices in the same device array created at built time or else this is going to be a can of worms.

Reply:
@tbursztyka  @dleach02 

> I quickly stopped when I realized this approach would create a schism in how devices are handled.

> That breaks too many things (parent/child ordering, thus all that relies on it such as PM etc...)

This is incorrect. The point of the multi API device model is to be interchangeable with the current API wrappers, PM, etc. The multi API device model should not break anything, if it does, please detail what is breaking so it can be handled.
The point of the multi api device model is to replace the current implementation.

There will only be two different ways of defining devices, the new one for which we should update the drivers and shells to use, and the current one, which will be kept to support drivers and apps which have not yet been updated to the new device model. The current one can then be deprecated when transition is done.

Regarding PM, using the runtime device PM, you can treat the device structs which belong to the same node as independent devices (which they aren't of course, but the application will not and won't need to know that).

Using the PM action, calling it on one or the other will result in the same action taken by the device, it will not break anything, it just means the user must be conscious of the fact that it is the same device they are interacting with when manually changing power states. 

The elegant solution to this in the future is to create a PM device API which devices which support PM can implement, moving the pm context from the device struct to the drivers themselves. This is another cool feature of having multiple APIs, no bloating the device struct context with potentially unused sub contexts like pm.

> Only for sake of multi-API use case.

The multi API device model fixes many issues, and as stated before, is a complete replacement for the current device model. It supporting multiple APIs is a necessary feature the current model lacks, but it also brings other features and cleanup:

Take for example the shells: We can now create a list at compile time containing every device which supports a specific API by node_id, and get additional properties defined for that device using the new properties files (see zephyr/samples/multi_Api_shell/sensor_shell/src/main.c). 

And the sensors: We can now define vendor, model and other properties which can be fetched at compile time, which is a good mechanism to fix https://github.com/zephyrproject-rtos/zephyr/pull/49294 for example.

The devicetree can be cleaned up since it no longer needs to generate extern declarations for devices since this is now handled by the new properties files and scripts.

>  now we would have devices existing in an array depending of their primary API, fully separated to the others

There are two lists of device structs in ROM, one which holds the current DEVICE_DEFINE devices, and one which holds all device structs created using DEVICE_API_DEFINE. The reason for this is that device_get_binding(), which is not necessary with the new multi api device model, iterates through that section. All devices defined using DEVICE_API_DEFINE are placed in a single section in ROM. They are not iterated through by design.

The point of the new API model is to allow it to coexist alongside the current device model, without altering it.

> Perhaps you should attack the problem from another stand point: the api property itself, leaving out of the picture how the device is treated.

This was one of my first proposed solutions to the issue, which was shut down pretty much immediately, and for good reason. Runtime lookup of APIs is slow and cumbersome, the multi API device model handles everything at compile time, completely removing the need for runtime lookup of devices in general as stated earlier (say goodbye device_get_binding :))

> 
```
static inline uint32_t z_impl_counter_get_frequency(const struct device *dev)
{
        const struct counter_driver_api *api = API_GET(dev, counter);
```

The problem here is the runtime lookup of the API, as stated previously, this is not feasible. It would also break every Zephyr API, which is not realistic, and as stated earlier "The multi API device model should not break anything"

I hope you will find the time to review the proposal in more detail, if not, i will go through it in the dev meeting thursday september 22.

Reply:
As stated in the quite cumbersome response comment, the device array is not relevant with the new design. The devices are all in the same device array in any case, just not the one which holds the devices defined with the current device model. 

No can of works here is to be found here :)